cmd/minimega/vm.go

```diff
@@ -314,6 +322,105 @@ func (vm *BaseVM) GetNetworks() []NetConfig {
 	return n
 }
 
+func (vm *BaseVM) GetBonds() map[string]NetConfigs {
+	return vm.Bonds
+}
+
+func (vm *BaseVM) AddBond(name string, ifaces []int, mode, lacp string, fallback, qinq bool) error {
+	vm.lock.Lock()
+	defer vm.lock.Unlock()
+
+	var (
+		netConfigs = make([]NetConfig, len(ifaces))
+		ifaceMap   = make(map[string]int) // name --> VLAN mapping
+		vlan       = -1
+
+		bridge string
+	)
+
+	for i, idx := range ifaces {
+		cfg, err := vm.getNetwork(idx)
+		if err != nil {
+			return fmt.Errorf("interface %d not found for vm %s", idx, vm.Name)
+		}
+
+		if bridge == "" {
+			bridge = cfg.Bridge
+		} else if cfg.Bridge != bridge {
+			return fmt.Errorf("interfaces being bonded are not on the same bridge")
+		}
+
+		if vlan < 0 {
+			vlan = cfg.VLAN
+		} else if cfg.VLAN != vlan {
+			log.Warn("interface %d on vm %s is not on VLAN %d -- still defaulting to %d for bond", idx, vm.Name, vlan, vlan)
+		}
+
+		qinq = qinq || cfg.QinQ
+
+		netConfigs[i] = cfg
+		ifaceMap[cfg.Tap] = cfg.VLAN
+	}
+
+	br, err := getBridge(bridge)
+	if err != nil {
+		return err
+	}
+
+	if name == "" {
+		name = br.CreateBondName()
+	}
+
+	if err := br.AddBond(name, mode, lacp, fallback, ifaceMap, vlan); err != nil {
+		return err
+	}
+
+	if qinq {
+		br.SetTapQinQ(name, vlan)
+	}
+
+	vm.Bonds[name] = netConfigs
+	return nil
+}
+
+func (vm *BaseVM) ClearBond(name string) error {
+	vm.lock.Lock()
+	defer vm.lock.Unlock()
+
+	ifaces, ok := vm.Bonds[name]
+	if !ok {
+		return fmt.Errorf("bond %s doesn't exist on VM %s", name, vm.Name)
+	}
+
+	if len(ifaces) == 0 {
+		delete(vm.Bonds, name)
+		return nil
+	}
+
+	bridge := ifaces[0].Bridge
+	br, err := getBridge(bridge)
+	if err != nil {
+		return err
+	}
+
+	if err := br.DeleteBond(name); err != nil {
+		return err
+	}
+
+	for _, c := range ifaces {
+		if c.QinQ {
+			// TODO: Should we ignore errors here? Or should we return immediately if
+			// an error occurs? Or should we gather up multiple errors, go ahead and
+			// delete the bond from "vm.Bonds", and then return the multiple errors as
+			// a single error?
+			br.SetTapQinQ(c.Tap, c.VLAN)
```

Comment:
we should try to cleanup if we can and return single error, and/or show as warning with the collection being an error?

Reply:
Okay @aherna can do. I'll work on that now.